cmake/FindBpfProg.cmake

```diff
@@ -0,0 +1,106 @@
+if(${CLANG_version_result} EQUAL 0)
+  if("${CLANG_version_output}" MATCHES "^clang version ([^\n]+)\n")
+    # Transform X.Y.Z into X;Y;Z which can then be interpreted as a list
+    set(CLANG_VERSION "${CMAKE_MATCH_1}")
+    string(REPLACE "." ";" CLANG_VERSION_LIST ${CLANG_VERSION})
+    list(GET CLANG_VERSION_LIST 0 CLANG_VERSION_MAJOR)
+
+    # Anything older than clang 10 doesn't really work
+    string(COMPARE LESS ${CLANG_VERSION_MAJOR} 10 CLANG_VERSION_MAJOR_LT10)
+    if(${CLANG_VERSION_MAJOR_LT10})
+      message(FATAL_ERROR "clang ${CLANG_VERSION} is too old for CO-RE bpf")
+    endif()
+
+    message(STATUS "Found clang version: ${CLANG_VERSION}")
+  else()
+    message(FATAL_ERROR "Failed to parse clang version string: ${CLANG_version_result}")
+  endif()
+else()
+  message(FATAL_ERROR "Command \"${CLANG_EXE} --version\" failed with output:\n${CLANG_version_error}")
+endif()
+
+# Generate vmlinux.h
+set(GENERATED_VMLINUX_DIR ${CMAKE_CURRENT_BINARY_DIR})
+set(GENERATED_VMLINUX ${GENERATED_VMLINUX_DIR}/vmlinux.h)
+execute_process(COMMAND ${BPFTOOL_EXE} btf dump file /sys/kernel/btf/vmlinux format c
+  OUTPUT_FILE ${GENERATED_VMLINUX}
+  ERROR_VARIABLE VMLINUX_error
+  RESULT_VARIABLE VMLINUX_result)
+if(${VMLINUX_result} EQUAL 0)
+  set(VMLINUX ${GENERATED_VMLINUX})
+else()
+  message(FATAL_ERROR "Failed to dump vmlinux.h from BTF: ${VMLINUX_error}")
+endif()
+
+# Get clang bpf system includes
+execute_process(COMMAND bash -c "${CLANG_EXE} -v -E - < /dev/null 2>&1 | sed -n '/<...> search starts here:/,/End of search list./{ s| \\(/.*\\)|-idirafter \\1|p }'"
+  OUTPUT_VARIABLE CLANG_SYSTEM_INCLUDES_output
+  ERROR_VARIABLE CLANG_SYSTEM_INCLUDES_error
+  RESULT_VARIABLE CLANG_SYSTEM_INCLUDES_result
+  OUTPUT_STRIP_TRAILING_WHITESPACE)
+if(${CLANG_SYSTEM_INCLUDES_result} EQUAL 0)
+  string(REPLACE "\n" " " CLANG_SYSTEM_INCLUDES ${CLANG_SYSTEM_INCLUDES_output})
+  message(STATUS "BPF system include flags: ${CLANG_SYSTEM_INCLUDES}")
+else()
+  message(FATAL_ERROR "Failed to determine BPF system includes: ${CLANG_SYSTEM_INCLUDES_error}")
+endif()
+
+# Get target arch
+execute_process(COMMAND uname -m
+  COMMAND sed "s/x86_64/x86/"
+  OUTPUT_VARIABLE ARCH_output
+  ERROR_VARIABLE ARCH_error
+  RESULT_VARIABLE ARCH_result
+  OUTPUT_STRIP_TRAILING_WHITESPACE)
+if(${ARCH_result} EQUAL 0)
+  set(ARCH ${ARCH_output})
+  message(STATUS "BPF target arch: ${ARCH}")
+else()
+  message(FATAL_ERROR "Failed to determine target architecture: ${ARCH_error}")
+endif()
+
+# Public macro
+macro(bpf_prog name input)
+  set(BPF_C_FILE ${CMAKE_CURRENT_SOURCE_DIR}/${input})
+  set(BPF_O_FILE ${CMAKE_CURRENT_BINARY_DIR}/${name}.bpf.o)
+  set(BPF_SKEL_FILE ${CMAKE_CURRENT_BINARY_DIR}/${name}.skel.h)
+  set(OUTPUT_TARGET ${name}_prog)
+
+  # Build BPF object file
+  add_custom_command(OUTPUT ${BPF_O_FILE}
+    COMMAND ${CLANG_EXE} -g -O2 -target bpf -D__TARGET_ARCH_${ARCH} ${CLANG_SYSTEM_INCLUDES} -I${GENERATED_VMLINUX_DIR} -c ${BPF_C_FILE} -o ${BPF_O_FILE}
```

Comment:
The current code does not use the libbpf submodule or the included bpftool. Should be possible to have magic variables for it, though. Just pointing it out.

Reply:
Now it does